engine/runtime-integration-tests/src/test/scala/org/enso/interpreter/test/instrument/RuntimeAsyncCommandsTest.scala

```diff
@@ -242,4 +243,93 @@ class RuntimeAsyncCommandsTest
     diagnostic.message shouldEqual Some("sleep interrupted")
     diagnostic.stack should not be empty
   }
+
+  it should "interrupt running execution context and recompute" in {
+    val moduleName = "Enso_Test.Test.Main"
+    val contextId  = UUID.randomUUID()
+    val requestId  = UUID.randomUUID()
+
+    val metadata = new Metadata
+    metadata.addItem(194, 7)
+    val code =
+      """from Standard.Base import all
+        |polyglot java import java.lang.Thread
+        |
+        |loop n s=0 =
+        |    if (s > n) then s else
+        |        Thread.sleep 100
+        |        loop n s+1
+        |
+        |main =
+        |    IO.println "started"
+        |    v = loop 50
+        |    v
+        |""".stripMargin.linesIterator.mkString("\n")
+    val contents = metadata.appendToCode(code)
+    val mainFile = context.writeMain(contents)
+
+    // create context
+    context.send(Api.Request(requestId, Api.CreateContextRequest(contextId)))
+    context.receive shouldEqual Some(
+      Api.Response(requestId, Api.CreateContextResponse(contextId))
+    )
+
+    // Open the new file
+    context.send(
+      Api.Request(requestId, Api.OpenFileRequest(mainFile, contents))
+    )
+    context.receive shouldEqual Some(
+      Api.Response(Some(requestId), Api.OpenFileResponse)
+    )
+
+    // push main
+    val item1 = Api.StackItem.ExplicitCall(
+      Api.MethodPointer(moduleName, "Enso_Test.Test.Main", "main"),
+      None,
+      Vector()
+    )
+    context.send(
+      Api.Request(requestId, Api.PushContextRequest(contextId, item1))
+    )
+    context.receiveNIgnoreExpressionUpdates(
+      1
+    ) should contain theSameElementsAs Seq(
+      Api.Response(requestId, Api.PushContextResponse(contextId))
+    )
+
+    // wait for program to start
+    var isProgramStarted = false
+    var iteration        = 0
+    while (!isProgramStarted && iteration < 100) {
+      val out = context.consumeOut
+      Thread.sleep(200)
+      isProgramStarted = out == List("started")
+      iteration += 1
+    }
+    if (!isProgramStarted) {
+      fail("Program start timed out")
+    }
+
+    // recompute/interrupt
+    context.send(
+      Api.Request(requestId, Api.RecomputeContextRequest(contextId, None, None))
+    )
+    val responses = context.receiveN(
+      4
+    )
+    responses should contain allOf (
+      Api.Response(requestId, Api.RecomputeContextResponse(contextId)),
+      context.executionComplete(contextId)
+    )
+    val updatesWithInterruptPanic = responses
+      .collect {
+        case response if response.payload.isInstanceOf[Api.ExpressionUpdates] =>
+          response.payload.asInstanceOf[Api.ExpressionUpdates]
+      }
+      .flatMap(_.updates)
+      .map(_.payload)
+      .filter(_.isInstanceOf[Payload.Panic])
+
+    updatesWithInterruptPanic.length shouldEqual 0
```

Comment:
Instead of looking for panic updates, I'd just specify all 4 expected messages with `theSameElementsAs`. I assume all the messages are known because the test waits for `executionComplete` anyway.